syz-verifier/main.go

```diff
@@ -242,20 +252,84 @@ func startRPCServer(vrf *Verifier) (*RPCServer, error) {
 }
 
 // Connect notifies the RPCServer that a new Runner was started.
-func (srv *RPCServer) Connect(a *rpctype.RunnerConnectArgs, r *int) error {
+func (srv *RPCServer) Connect(a *rpctype.RunnerConnectArgs, r *rpctype.RunnerConnectRes) error {
 	srv.mu.Lock()
 	defer srv.mu.Unlock()
 	pool, vm := a.Pool, a.VM
 	srv.pools[pool].vmRunners[vm] = nil
+	if !srv.pools[pool].checked {
+		// This is the first VM connecting for this kernel so we need to find
+		// the list of unsupported system calls.
+		r.CheckUnsupportedCalls = true
+	}
+	return nil
+}
+
+// CheckUnsupported communicates to the server the list of system calls not
+// supported by the kernel corresponding to this pool and updates the list of
+// enabled system calls. This function is called once for each kernel.
+// When all kernels have reported the list of unsupported system calls, the
+// choice table will be created using only the system calls supported by all
+// kernels.
+func (srv *RPCServer) CheckUnsupported(a *rpctype.CheckUnsupportedArgs, r *int) error {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	if srv.pools[a.Pool].checked {
+		return nil
+	}
+	srv.pools[a.Pool].checked = true
+	vrf := srv.vrf
+
+	for _, unsupported := range a.UnsupportedCalls {
+		for c := range vrf.calls {
+			if c.ID == unsupported.ID {
+				vrf.reasons[c] = unsupported.Reason
+				break
+			}
+		}
+	}
+
+	srv.notChecked--
+	if srv.notChecked == 0 {
+		vrf.finalizeCallSet(os.Stdout)
+		vrf.choiceTable = vrf.target.BuildChoiceTable(nil, vrf.calls)
+		srv.cond.Signal()
+	}
 	return nil
 }
 
+// finalizeCallSet removes the system calls that are not supported from the set
+// of enabled system calls and reports the reason to the io.Writer (either
+// because the call is not supported by one of the kernels or because the call
+// is missing some transitive dependencies). The resulting set of system calls
+// will be used to build the prog.ChoiceTable.
+func (vrf *Verifier) finalizeCallSet(w io.Writer) {
+	if len(vrf.reasons) > 0 {
+		fmt.Fprintln(w, "Calls not supported by kernels:")
+	}
+	for c, reason := range vrf.reasons {
+		fmt.Fprintf(w, "disabling syscall %v: %v\n", c.Name, reason)
```

Comment:
This may produce thousands of lines of output if run with all syscalls, but user did not really asked to enabled anything specifically, so most likely they are not interested in any concrete syscall.
I think it's better to mimic syz-manger logic for printing disabled syscalls.

Reply:
Since we already printed a header line `Calls not supported by kernels:`, it does not make sense to print `disabling syscall` on every line. It only takes space. So I would either remove the header line, or replace `disabling syscall` with 2 spaces or something.

Reply:
I meant not printing messages about disabled calls if user did not ask to enable anything specifically (no "enable_syscalls" in config).

Reply:
Ah, I see.